datafusion/optimizer/src/extract_equijoin_predicate.rs

```diff
@@ -0,0 +1,438 @@
+    fn try_optimize(
+        &self,
+        plan: &LogicalPlan,
+        config: &dyn OptimizerConfig,
+    ) -> Result<Option<LogicalPlan>> {
+        match plan {
+            LogicalPlan::Join(Join {
+                left,
+                right,
+                on,
+                filter,
+                join_type,
+                join_constraint,
+                schema,
+                null_equals_null,
+            }) => {
+                let left_schema = left.schema();
+                let right_schema = right.schema();
+
+                let new_on_and_accumu_filter = if let Some(expr) = filter {
+                    let mut accum: Vec<(Expr, Expr)> = vec![];
+                    let mut accum_filter: Vec<Expr> = vec![];
+                    // TODO: avoding clone with split_conjunction
+                    extract_join_keys(
+                        expr.clone(),
+                        &mut accum,
+                        &mut accum_filter,
+                        left_schema,
+                        right_schema,
+                    )?;
+
+                    (!accum.is_empty()).then(|| {
+                        let mut new_on = on.clone();
+                        new_on.extend(accum);
+                        (new_on, accum_filter)
+                    })
+                } else {
+                    None
+                };
+
+                let optimized_left = self.try_optimize(left.as_ref(), config)?;
+                let optimized_right = self.try_optimize(right.as_ref(), config)?;
+                let plan_changed = new_on_and_accumu_filter.is_some()
+                    || optimized_left.is_some()
+                    || optimized_right.is_some();
+
+                let plan = plan_changed.then(|| {
+                    let left =
+                        optimized_left.map(Arc::new).unwrap_or_else(|| left.clone());
+                    let right = optimized_right
+                        .map(Arc::new)
+                        .unwrap_or_else(|| right.clone());
+                    let (new_on, new_filter) = new_on_and_accumu_filter
+                        .map(|(on, accumu_filter)| {
+                            let filter = accumu_filter.into_iter().reduce(Expr::and);
+                            (on, filter)
+                        })
+                        .unwrap_or_else(|| (on.clone(), filter.clone()));
+
+                    LogicalPlan::Join(Join {
+                        left,
+                        right,
+                        on: new_on,
+                        filter: new_filter,
+                        join_type: *join_type,
+                        join_constraint: *join_constraint,
+                        schema: schema.clone(),
+                        null_equals_null: *null_equals_null,
+                    })
+                });
+
+                Ok(plan)
+            }
+            _ => Ok(Some(optimize_children(self, plan, config)?)),
```

Comment:
@jackwener has been removing direct recursion in optimizer rules -- for example see https://github.com/apache/arrow-datafusion/pull/4687

Perhaps we could do the same with this rule too

Reply:
Yes, we can do it.

Utilizing the traverse of optimizer now.